js/src/SearchBar.js

```diff
@@ -1,158 +1,171 @@
+class SearchBar extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      visible: props.visible,
+      current: 1,
+      searchItem: null,
+      counter: ''
+    }
+  }
 
-    this.selection = container;
-    this.map = map;
-    this.search_index = search_index;
+  componentWillReceiveProps (nextProps) {
+    this.props.map.key_manager.add_escape_listener(() => this.close(), true)
+    this.setState({
+      ...nextProps,
+      current: 1,
+      results: null,
+      searchItem: null,
+      counter: '',
+      clearNext: this.props.map.key_manager.add_key_listener(
+        ['enter', 'ctrl+g'], () => this.next(), false),
+      clearPrevious: this.props.map.key_manager.add_key_listener(
+        ['shift+enter', 'shift+ctrl+g'], () => this.previous(), false)
+    })
+  }
 
-    this.current = 1;
-    this.results = null;
+  componentDidUpdate () {
+    this.inputRef.focus()
+  }
 
-    var on_input_fn = function(input) {
-        this.current = 1;
-        this.results = _drop_duplicates(this.search_index.find(input.value));
-        this.update();
-    }.bind(this, this.input.node());
-    this.input.on('input', utils.debounce(on_input_fn, 200));
-}
-
-var comp_keys = {
-    metabolite: ['m', 'node_id'],
-    reaction: ['r', 'reaction_id'],
-    text_label: ['t', 'text_label_id']
-};
-function _drop_duplicates(results) {
-    return _.uniq(results, function(item) {
-        // make a string for fast comparison
-        var t = comp_keys[item.type];
-        return t[0] + item[t[1]];
-    });
-}
+  handleInput (value) {
+    if (this.state.visible) {
+      const results = this.dropDuplicates(this.props.map.search_index.find(value))
+      let counter = ''
+      if (results === null || value === '' || !value) {
+        counter = ''
+        this.props.map.highlight(null)
+      } else if (results.length === 0) {
+        counter = '0 / 0'
+        this.props.map.highlight(null)
+      } else {
+        counter = (this.state.current + ' / ' + results.length)
+        var r = results[this.state.current - 1]
+        if (r.type === 'reaction') {
+          this.props.map.zoom_to_reaction(r.reaction_id)
+          this.props.map.highlight_reaction(r.reaction_id)
+        } else if (r.type === 'metabolite') {
+          this.props.map.zoom_to_node(r.node_id)
+          this.props.map.highlight_node(r.node_id)
+        } else if (r.type === 'text_label') {
+          this.props.map.zoom_to_text_label(r.text_label_id)
+          this.props.map.highlight_text_label(r.text_label_id)
+        } else {
+          throw new Error('Bad search index data type: ' + r.type)
+        }
+      }
+      this.setState({
+        searchItem: value,
+        current: 1,
+        counter,
+        results
+      })
+    }
+  }
 
-function is_visible() {
-    return this.selection.style('display') !== 'none';
-}
+  dropDuplicates (results) {
+    const compKeys = {
+      metabolite: ['m', 'node_id'],
```

Comment:
IMO this would be more readable as follows:
```
dropDuplicates (results) {
  const compKeys = {
    metabolite: {
      type: 'm',
      key: 'node_id'
    },
    reaction: {
      type: 'r',
      key: 'reaction_id'
    },
    text_label: {
      type: 't',
      key: 'text_label_id'
    }
  }
  return _.uniq(results, item => {
    // make a string for fast comparison
    const {type, key} = compKeys[item.type]
    return `${type}${item[key]}`
  })
}
```

Reply:
Done